### PR TITLE
refactor onoff to separate async notification and standardize terminology

### PR DESCRIPTION
**Pull request by pabigot**
This PR separates generic resource management updates that appear to be nearly stable from #23229.  The changes are:

* refactors the onoff service to reduce space by putting static configuration data into a separate structure;
* extracts the async notification infrastructure from the on-off service into a self-contained structure that can be used anywhere asynchronous operations may be used;
* recasts the on-off service as a manager that is to be used by services (including API changes to make it more consistent with other managers to come);

### CODE REVIEW

File: lib/os/onoff.c [hunk 1]
```diff
@@ -1,127 +1,81 @@
...
-		break;
-#ifdef CONFIG_POLL
-	case ONOFF_CLIENT_NOTIFY_SIGNAL:
-		k_poll_signal_raise(cli->async.signal, res);
-		break;
-#endif /* CONFIG_POLL */
-	default:
-		__ASSERT_NO_MSG(false);
+	if (cb) {
+		cb(mgr, cli, cli->user_data, res);
```
Comment (nordic-krch) [inline lib/os/onoff.c]: there was a concern that `user_data` might be overwritten by this time because of fetch result reports that client is completed after returning from `async_notify_finalize`. You mentioned that it will be documented https://github.com/zephyrproject-rtos/zephyr/pull/23229#discussion_r391646030. Can you point me to that? Anyway, i think that it would be safer to store `user_data` before calling `async_notify_finalize`

**Review by nordic-krch — COMMENTED** [review]
few comments.

**Comment by pabigot** [comment]
To proactively address some comments from earlier reviews for which I have not made changes:

* Yes, `struct async_notify` records the result of the operation in a common field rather than in the union wjere it would be available only for spin-wait notifications.  An argument can be made that this is unnecessary, since for signal the result is stored in `struct k_poll_signal` and for callback it's passed as an argument, thus only spin-wait needs it.  Reasons for retaining the existing approach include:
  * Consistency: the result of the operation can always be obtained using `async_notify_fetch_result()` regardless of the notification method used;
  * Flexibility: An application may multiplex notifications onto one `k_poll_signal` object (e.g. fire off a dozen and wake up when the first one completes; tricky but not inconceivable), or may not process the operation completion in the callback (in which case the result needs to be persisted somewhere, most likely on the client data that contains the notification result).  If the result is only retained for spinwait notifications the applications have to do extra work.

* Yes, the names are long.  They're also unambiguous and unlikely to conflict with future API.  If there is wide consensus that a specific alternative prefix is preferred I'll change them.  (`anotify_` instead of `async_notify_` has been proposed, but "async" means "not sync(hronous)" so "anotify" means "not notified"?  Meh.)

File: tests/lib/async_notify/src/main.c [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	rc = async_notify_validate(&notify);
+	zassert_equal(rc, 0,
+		      "init_spinwait invalid");
+
+	zassert_false(async_notify_uses_callback(&notify),
+		      "uses callback");
+
+	zassert_equal(notify.flags, ASYNC_NOTIFY_METHOD_SIGNAL,
+		      "flags mismatch");
+	zassert_equal(notify.method.signal, &sig,
```
[the remark this answers is not in the document]
Comment (pabigot) [reply tests/lib/async_notify/src/main.c]: "Need" no, but it does confirm that the implementation is doing what the implementer intended, even if that can be inferred from later behavior.  I prefer to verify the internal details where it's helpful to me as maintainer.

File: include/sys/async_notify.h [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+ * when a pointer to the object is passed to a service function that
+ * is documented to take control of the object, such as
+ * onoff_service_request().  While the service provider controls the
+ * object the client must not change any object fields.  Control
+ * reverts to the client:
+ * * if the call to the service API returns an error;
+ * * when operation completion is posted.  This may occur before the
+ *   call to the service API returns.
+ *
+ * Operation completion is technically posted when the flags field is
```
Comment (pabigot) [inline include/sys/async_notify.h]: @nordic-krch for https://github.com/zephyrproject-rtos/zephyr/pull/23601#discussion_r395066208 this is where the documentation was added.  I'll fix the spelling error in the next pass.

File: lib/os/onoff.c (hunk 1 above)
Comment (pabigot) [reply lib/os/onoff.c]: Done. For this API I can't see that it hurts, other than suggesting that doing this is sufficient to avoid the problem of repurposing the notification structure.  If somebody provides a callback that needs additional context from a containing object that won't have been cached.

**Review by nordic-krch — APPROVED** [review]
Could you fix 80liners checkpatch complains? Regarding `result` field in the async notify. Let's keep it for now as it is. I don't see clearly the use case that needs it so it's hard for me to say if there are other alternatives. If use case is specific to a service then first that comes to my mind is to implement that in the service, in completion callback (before calling `async_notify_finalize`) result can be stored in dedicated service struct which contains async notification. This way, it would be supported only for services which implements that but on the other hand it will save 4 bytes for each async notify which does not need it.

**Comment by pabigot** [comment]
> Could you fix 80liners checkpatch complains?

I think I got all but the ones where there's no obvious place to break a line.

**Comment by nordic-krch** [comment]
@tbursztyka @anangl if any of you could review that? @carlescufi maybe you know someone that would have interest in reviewing that?

File: include/sys/async_notify.h [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+ */
+
+/* Forward declaration */
+struct async_notify;
+
+/*
+ * Flag value that overwrites the method field when the operation has
+ * completed.
+ */
+#define ASYNC_NOTIFY_METHOD_COMPLETED 0
```
Comment (carlescufi) [inline include/sys/async_notify.h]: Again, I disagree with these long names. I really would like to shorten this to something like `ANOTIFY_` and `anotify_` or something even shorter, if we can come up with one. I will add this to the API meeting.

**Review by carlescufi — CHANGES_REQUESTED** [review]

**Comment by carlescufi** [comment]
> @tbursztyka @anangl if any of you could review that? @carlescufi maybe you know someone that would have interest in reviewing that?

No, unfortunately beyond @tbursztyka and @anangl I can't think of anyone else at this point. Perhaps @wentongwu or @pdunaj?

File: include/sys/async_notify.h (hunk 4 above)
Comment (tbursztyka) [reply include/sys/async_notify.h]: The whole namespace would need to change. I don't know if we need to keep async in the name, notify_* only perhaps?

File: lib/os/onoff.c [hunk 5]
```diff
@@ -1,39 +1,39 @@
...
+	 | ONOFF_RESET_SLEEPS)
 
 #define SERVICE_REFS_MAX UINT16_MAX
 
-#define SERVICE_STATE_OFF 0
-#define SERVICE_STATE_ON ONOFF_SERVICE_INTERNAL_BASE
-#define SERVICE_STATE_TRANSITION (ONOFF_SERVICE_INTERNAL_BASE << 1)
-#define SERVICE_STATE_TO_ON (SERVICE_STATE_TRANSITION | SERVICE_STATE_ON)
-#define SERVICE_STATE_TO_OFF (SERVICE_STATE_TRANSITION | SERVICE_STATE_OFF)
+#define ST_OFF 0
```
Comment (tbursztyka) [inline lib/os/onoff.c]: about naming, I don't know if going to the shortest possible one is relevant (it becomes cryptic then). You'll need to find something in the middle.

**Comment by tbursztyka** [comment]
besides naming issues, lgtm

File: include/sys/async_notify.h (hunk 4 above)
Comment (nordic-krch) [reply include/sys/async_notify.h]: in a way, by default notification is asynchronous in nature so `async_` prefix might be redundant. 
Some other proposals :
- `notifier`
- `gnotify` - generic notification
- `k_notify` - since (a least currently) it is limited to kernel.

Comment (tbursztyka) [reply include/sys/async_notify.h]: gnotify sounds like glib, your are right k_ is strictly used by the kernel only. z_notify maybe? Though it does not "sound" nice.

Comment (nordic-krch) [reply include/sys/async_notify.h]: yeah, i thought about z_notify but `z_` indicates internal function.

Comment (pabigot) [reply include/sys/async_notify.h]: I suppose that `async_` doesn't add a huge amount of information except that it's intended to be the standard way of notifying of completion of async operations.

Bare `notify_` IMO has too large of a conflict with existing and new identifiers that are not part of this API (`struct notify_data`, `notify_remote_info`).

Since this is in the `sys/` include hierarchy `sys_notify_` could mitigate that.

Comment (carlescufi) [reply include/sys/async_notify.h]: API meeting:
`sys_notify_` raised no objections.

Comment (nordic-krch) [reply include/sys/async_notify.h]: how to we call the file? `notify.h`? since path contains the namespace (`sys/notify.h`)?

Comment (tbursztyka) [reply include/sys/async_notify.h]: yes, as all the other files there. (besides sys_io.h which is a legacy name, and was not part of sys/ at first if I remember well)

File: doc/reference/misc/notify.rst [hunk 6]
```diff
@@ -0,0 +1,25 @@
+.. _async_notification:
```
Comment (nordic-krch) [inline doc/reference/misc/notify.rst]: this could be called `sys_notify` now to be aligned with `sys_notify_apis`

Comment (pabigot) [reply doc/reference/misc/notify.rst]: It can't: the target name `sys_notify` refers to `struct sys_notify`.

**Comment by nordic-krch** [comment]
@carlescufi @tbursztyka naming has been resolved can you look again?

**Review by carlescufi — APPROVED** [review]

File: include/sys/notify.h [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+extern "C" {
+#endif
+
+/**
+ * @defgroup sys_notify_apis Asynchronous Notification APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/* Forward declaration */
```
Comment (jukkar) [inline include/sys/notify.h:24]: This comment could be removed, it just states the obvious.

File: include/sys/notify.h [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+struct sys_notify;
+
+/*
```
Comment (jukkar) [inline include/sys/notify.h:27]: This looks like a comment that should be in the documentation, so we should put `/**` here.
Same comment to other defines below.
Could we replace these defines by enum as they seem to be sequential and related to each other?

File: include/sys/notify.h [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+#define SYS_NOTIFY_METHOD_SIGNAL 2
+
+/*
+ * Select notification through a user-provided callback.
+ *
+ * See sys_notify_init_callback().
+ */
+#define SYS_NOTIFY_METHOD_CALLBACK 3
+
+#define SYS_NOTIFY_METHOD_MASK 0x03
```
Comment (jukkar) [inline include/sys/notify.h:57]: If we do not want to document these, they should probably be outside of doxygen group or marked as internal using `/** @cond INTERNAL_HIDDEN */`

File: include/sys/notify.h [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+ *
+ * The content of this structure is not public API to clients: all
+ * configuration and inspection should be done with functions like
+ * sys_notify_init_callback() and sys_notify_fetch_result().
+ * However, services that use this structure may access certain
+ * fields directly.
+ */
+struct sys_notify {
+	union method {
+		/* Pointer to signal used to notify client.
```
Comment (jukkar) [inline include/sys/notify.h:141]: Documentation comment here `/**`

File: include/sys/notify.h [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+		 * The signal value corresponds to the res parameter
+		 * of sys_notify_callback.
+		 */
+		struct k_poll_signal *signal;
+
+		/* Generic callback function for callback notification. */
+		sys_notify_generic_callback callback;
+	} method;
+
+	/*
```
Comment (jukkar) [inline include/sys/notify.h:152]: ditto

File: include/sys/notify.h [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+	 *
+	 * Bits at and above SYS_NOTIFY_EXTENSION_POS are available
+	 * for use by service extensions while the containing object
+	 * is managed by the service.  They are not for client use,
+	 * are zeroed by the async notify API init functions, and will
+	 * be zeroed by sys_notify_finalize().
+	 */
+	u32_t volatile flags;
+
+	/*
```
Comment (jukkar) [inline include/sys/notify.h:168]: and here too

File: include/sys/notify.h [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+	 * The result of the operation.
+	 *
+	 * This is the value that was (or would be) passed to the
+	 * async infrastructure.  This field is the sole record of
+	 * success or failure for spin-wait synchronous operations.
+	 */
+	int volatile result;
+};
+
+/** @internal */
```
Comment (jukkar) [inline include/sys/notify.h:178]: In other parts of the code, I have used
```
/** @cond INTERNAL_HIDDEN */
...
/** @endcond */
```
not sure which is correct

File: include/sys/notify.h [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+{
+	__ASSERT_NO_MSG(notify != NULL);
+	__ASSERT_NO_MSG(result != NULL);
+	int rv = -EAGAIN;
+
+	if (sys_notify_get_method(notify) == SYS_NOTIFY_METHOD_COMPLETED) {
+		rv = 0;
+		*result = notify->result;
+	}
+	return rv;
```
Comment (jukkar) [inline include/sys/notify.h:249]: empty line after }

File: include/sys/notify.h [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+ *
+ * On completion of the operation the client object must be
+ * reinitialized before it can be re-used.
+ *
+ * @param notify pointer to the notification configuration object.
+ *
+ * @param handler a function pointer to use for notification.
+ */
+static inline void sys_notify_init_callback(struct sys_notify *notify,
+					      sys_notify_generic_callback handler)
```
Comment (jukkar) [inline include/sys/notify.h:324]: This looks like too long line, in this case I would just make sure the line is max 80 chars long even if the indentation would not match. Other option is just to put the 1st parameter to new line.

File: tests/lib/notify/testcase.yaml [hunk 16]
```diff
@@ -0,0 +1,3 @@
+tests:
+  libraries.sys_notify:
+    tags: sys_notify
```
Comment (jukkar) [inline tests/lib/notify/testcase.yaml:3]: You could also add here `notify` tag (easier to remember)

**Review by jukkar — COMMENTED** [review]
Some random notes

File: include/sys/notify.h (hunk 8 above)
Comment (pabigot) [reply include/sys/notify.h:20]: It's not public API, but maintainers need to know its intent so it's described.  It's also the value of a bitfield in flags so it shouldn't be an enum.

File: include/sys/notify.h (hunk 9 above)
Comment (pabigot) [reply include/sys/notify.h:50]: Moved.

File: include/sys/notify.h (hunk 10 above)
Comment (pabigot) [reply include/sys/notify.h:140]: It's not public API (it's manipulated through accessor functions).  And documentation for struct members isn't generated so it wouldn't help.

File: include/sys/notify.h (hunk 13 above)
Comment (pabigot) [reply include/sys/notify.h:177]: I'm not sure either work, but `@internal` is common.

File: include/sys/notify.h (hunk 15 above)
Comment (pabigot) [reply include/sys/notify.h:324]: Generally we don't do that sort of thing just to satisfy the line length test (which is advisory in Zephyr).

File: tests/lib/notify/testcase.yaml (hunk 16 above)
Comment (pabigot) [reply tests/lib/notify/testcase.yaml]: replaced

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by carlescufi** [comment]
> besides naming issues, lgtm

Will merge since the naming has been resolved
